src/NEPTypes.jl

```diff
@@ -179,122 +193,121 @@ julia> compute_Mder(nep,1)-(A0+A1*exp(1))
-                # with sparsity pattern of M() for optimization
-                Z=copy(nep.Zero)
-            else
-                Z=spzeros(eltype(V),size(V,1),size(V,2))
-            end
-        else
-            Z=zeros(eltype(V),size(V))
-        end
-        # Sum together all the terms in the SPMF:
-        if(nep.Schur_factorize_before) #Optimize if allowed to factorize before
-            (T, Q, ) = schur(S)
-        end
-        for i=1:size(nep.A,1)
+    function compute_MM(nep::SPMF_NEP{T,Ftype},S::AbstractMatrix,V::AbstractMatrix) where {T,Ftype}
+
+        AA=get_Av(nep);
+        ff=get_fv(nep);
+        m=size(ff,1); n=size(nep,1); p=size(S,1);
+
+        # Type logic including Ftype
+        FStype=promote_type(eltype(S),Ftype) # eltype of f(S)
+        T0=promote_type(eltype(V),eltype(AA[1]),FStype) # Output type
+
+
+        # Always return a dense matrix
+        Z=zeros(T0,n,p);
+
+        # Sum together all the terms in the SPMF: (temporarily disabled)
+        #if (nep.Schur_factorize_before) #Optimize if allowed to factorize before
+        #    (TT, Q, ) = schur(S)  # Currently not used
+        #end
+
+        VFi=Matrix{promote_type(FStype,eltype(V))}(undef,n,p); # Type of V*f(S)
+        local Fi=Matrix{FStype}(undef,p,p);
+        for i=1:m
             ## Compute Fi=f_i(S) in an optimized way
             if (isdiag(S)) # optimize if S is diagonal
                 Sd=diag(S);
+                local Fid::Vector{FStype}
                 if (norm(Sd .- Sd[1])==0) # Optimize further if S is a multiple of identity
-                    Fid=nep.fi[i](reshape([Sd[1]],1,1))[1]*ones(size(Sd,1))
+                    Fid=nep.fi[i](Sd[1])*ones(T0,p)
                 else  # Diagonal but not constant
-                    Fid=zeros(ComplexF64,size(S,1))
-                    for j=1:size(S,1)
-                        Fid[j]=nep.fi[i](reshape([Sd[j]],1,1))[1]
+                    Fid0=Vector{Number}(undef,p)
+                    for j=1:p
+                        Fid0[j]=nep.fi[i](Sd[j])
                     end
+                    Fid=Vector{FStype}(Fid0);
                 end
-                Fi=sparse(Diagonal(Fid))
-            else  # Otherwise just compute the matrix function operation
-                if(nep.Schur_factorize_before)
-                    Fi= Q*nep.fi[i](T)*Q'
-                else
-                    Fi=nep.fi[i](S)
-                end
-            end
-            ## Sum it up
-
-
-            VFi=V*Fi;
-            if (isa(nep.A[i],SubArray) && (eltype(nep.A[i]) != eltype(VFi)))
-                # 2018-09-14: SubArray x Matrix of different types do not work? Check in new version of Julia/base at later point (TODO)
-                # https://discourse.julialang.org/t/subarray-x-matrix-multipliciation-of-different-eltype-fails/14950
-                # Workaround by making a matrix copy.
-                Z=Z+copy(nep.A[i])*VFi;
+                Fi[:,:]=Diagonal(Fid)
```

Comment:
Could we use fusing dot here instead of `[:,:]`? I.e. `Fi .= Diagonal(Fid)`. And similarly below. A quick benchmark suggests that fusing dot is a lot faster, but I'm not so familiar with the colon syntax, perhaps it does something else?

Reply:
I agree.